demos/adaptive_multigrid/demo.py.rst

```diff
@@ -0,0 +1,195 @@
+      v = TestFunction(V)
+      bc = DirichletBC(V, Constant(0), "on_boundary")
+      f = Constant(1)
+      F = inner(grad(uh), grad(v))*dx - inner(f, v)*dx
+
+      problem = NonlinearVariationalProblem(F, uh, bc)
+      solver = NonlinearVariationalSolver(problem, solver_parameters=params)
+      solver.set_transfer_manager(atm)
+      solver.solve()
+      return uh
+
+Note the code after the construction of the NonlinearVariationalProblem(). To use the AdaptiveMeshHierarchy with the existing Firedrake solver, we have to set the AdaptiveTransferManager as the transfer manager of the multigrid solver.
+For the parameters of the multigrid solver, we will be using patch relaxation, which we define with ::
+   solver_params = {
+            "mat_type": "matfree",
+            "ksp_type": "cg",
+            "pc_type": "mg",
+            "mg_levels": {
+                "ksp_type": "chebyshev",
+                "ksp_max_it": 1,
+                "pc_type": "python",
+                "pc_python_type": "firedrake.PatchPC",
+                "patch": {
+                    "pc_patch": {
+                        "construct_type": "star",
+                        "construct_dim": 0,
+                        "sub_mat_type": "seqdense",
+                        "dense_inverse": True,
+                        "save_operators": True,
+                        "precompute_element_tensors": True,
+                    },
+                    "sub_ksp_type": "preonly",
+                    "sub_pc_type": "lu",
+                },
+            },
+            "mg_coarse": {
+                "ksp_type": "preonly",
+                "pc_type": "python",
+                "pc_python_type": "firedrake.AssembledPC",
+                "assembled": {"ksp_type": "preonly", "pc_type": "lu"},
+            },
+        }
+
+For more information about patch relaxation, see `Using patch relaxation for multigrid <https://www.firedrakeproject.org/demos/poisson_mg_patches.py.html>`_. The initial solution is shown below
+
+.. figure:: solution_l1.png
+   :align: center
+   :alt: Initial Solution from multigrid with initial mesh.
+
+
+Adaptive Mesh Refinement
+-------------------------
+In this section we will discuss how to adaptively refine select elements and add the newly refined mesh into the AdaptiveMeshHierarchy.
+For this problem, we will be using the Babuška-Rheinbolt a-posteriori estimate for an element:
+
+.. math::
+   \eta_K^2 = h_K^2 \int_K | f + \nabla^2 u_h |^2 \mathrm{d}x + \frac{h_K}{2} \int_{\partial K \setminus \partial \Omega} \llbracket \nabla u_h \cdot n \rrbracket^2 \mathrm{d}s,
+
+where :math:`K` is the element, :math:`h_K` is the diameter of the element, :math:`n` is the normal, and :math:`\llbracket \cdot \rrbracket` is the jump operator. The a-posteriori estimator is computed using the solution at the current level :math:`h`. We can use a trick to compute the estimator on each element. We transform the above estimator into the variational problem 
```

Comment:
Say something like "Integrating over the domain and using the fact that the functions are piecewise constant on each cell, we transform ..."